paperqa/agents/search.py

```diff
@@ -0,0 +1,486 @@
+
+PQA_INDEX_ABSOLUTE_PATHS = (
+    os.environ.get("PQA_INDEX_ABSOLUTE_PATHS", "true").lower() == "true"
+)
+
+
+class AsyncRetryError(Exception):
+    pass
+
+
+class RobustEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, UUID):
+            # if the obj is uuid, we simply return the value of uuid
+            return str(obj)
+        if isinstance(obj, set):
+            return list(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+class SearchDocumentStorage(str, Enum):
+    JSON_MODEL_DUMP = "json_model_dump"
+    PICKLE_COMPRESSED = "pickle_compressed"
+    PICKLE_UNCOMPRESSED = "pickle_uncompressed"
+
+    def extension(self) -> str:
+        if self == SearchDocumentStorage.JSON_MODEL_DUMP:
+            return "json"
+        if self == SearchDocumentStorage.PICKLE_COMPRESSED:
+            return "zip"
+        return "pkl"
+
+    def write_to_string(self, data: BaseModel | SupportsPickle) -> bytes:
+        if self == SearchDocumentStorage.JSON_MODEL_DUMP:
+            if isinstance(data, BaseModel):
+                return json.dumps(data.model_dump(), cls=RobustEncoder).encode("utf-8")
+            raise ValueError("JSON_MODEL_DUMP requires a BaseModel object.")
+        if self == SearchDocumentStorage.PICKLE_COMPRESSED:
+            return zlib.compress(pickle.dumps(data))
+        return pickle.dumps(data)
+
+    def read_from_string(self, data: str | bytes) -> BaseModel | SupportsPickle:
+        if self == SearchDocumentStorage.JSON_MODEL_DUMP:
+            return json.loads(data)
+        if self == SearchDocumentStorage.PICKLE_COMPRESSED:
+            return pickle.loads(zlib.decompress(data))  # type: ignore[arg-type] # noqa: S301
+        return pickle.loads(data)  # type: ignore[arg-type] # noqa: S301
+
+
+class SearchIndex:
+
+    REQUIRED_FIELDS: ClassVar[list[str]] = ["file_location", "body"]
+
+    def __init__(
+        self,
+        fields: list[str] | None = None,
+        index_name: str = "pqa_index",
+        index_directory: str | os.PathLike | None = None,
+        storage: SearchDocumentStorage = SearchDocumentStorage.PICKLE_COMPRESSED,
+    ):
+        if fields is None:
+            fields = ["file_location", "body"]
+        self.fields = fields
+        if not all(f in self.fields for f in self.REQUIRED_FIELDS):
+            raise ValueError(
+                f"{self.REQUIRED_FIELDS} must be included in search index fields."
+            )
+        if index_directory is None:
+            index_directory = pqa_directory("indexes")
+        self.index_name = index_name
+        self._index_directory = index_directory
+        self._schema = None
+        self._index = None
+        self._searcher = None
+        self._index_files: dict[str, str] = {}
+        self.changed = False
+        self.storage = storage
+
+    async def init_directory(self):
+        await anyio.Path(await self.index_directory).mkdir(parents=True, exist_ok=True)
```

Comment:
Why are we using `anyio.Path`? Can we just use `pathlib.Path`?

Reply:
Talked about this live, but anyio's Path is async so reads/writes can occur in worker threads.